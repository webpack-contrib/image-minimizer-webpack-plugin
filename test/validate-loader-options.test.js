import path from "path";

import ImageMinimizerPlugin from "../src";

import { fixturesPath, plugins, webpack } from "./helpers";

describe("validate loader options", () => {
  const tests = {
    minify: {
      success: [
        ImageMinimizerPlugin.imageminMinify,
        [ImageMinimizerPlugin.imageminMinify],
      ],
      failure: [1, true, false, null],
    },
    minimizerOptions: {
      success: [
        { plugins: [] },
        {},
        {
          filename: "[name].[ext]",
        },
        {
          filename: () => "[name].[ext]",
        },
        {
          deleteOriginal: true,
        },
        {
          filter: () => {},
        },
      ],
      failure: [
        1,
        true,
        false,
        [],
        null,
        {
          filename: [],
        },
        {
          filename: {},
        },
        {
          filename: true,
        },
        {
          deleteOriginal: {},
        },
        {
          deleteOriginal: [],
        },
        {
          deleteOriginal: () => {},
        },
        {
          filter: 1,
        },
        {
          filter: true,
        },
        {
          filter: {},
        },
        {
          filter: [],
        },
        {
          filter: null,
        },
      ],
    },
    severityError: {
      success: ["error"],
      failure: [true, false, {}, [], () => {}],
    },
    unknown: {
      success: [],
      failure: [1, true, false, "test", /test/, [], {}, { foo: "bar" }],
    },
  };

  function stringifyValue(value) {
    if (
      Array.isArray(value) ||
      (value && typeof value === "object" && value.constructor === Object)
    ) {
      return JSON.stringify(value);
    }

    return value;
  }

  function createTestCase(key, value, type) {
    it(`should ${
      type === "success" ? "successfully validate" : "throw an error on"
    } the "${key}" option with "${stringifyValue(value)}" value`, async () => {
      const options = {
        entry: path.join(fixturesPath, "validate-options.js"),
        imageminLoaderOptions: {
          [key]: value,
        },
      };

      if (key === "severityError") {
        options.imageminLoaderOptions.minimizerOptions = { plugins };
      }

      let stats;

      try {
        stats = await webpack(options);
      } finally {
        const shouldSuccess = type === "success";
        const {
          compilation: { errors },
        } = stats;

        expect(stats.hasErrors()).toBe(!shouldSuccess);
        expect(errors).toHaveLength(shouldSuccess ? 0 : 1);

        if (!shouldSuccess) {
          // eslint-disable-next-line jest/no-conditional-expect
          expect(() => {
            throw new Error(errors[0].error.message);
          }).toThrowErrorMatchingSnapshot();
        }
      }
    });
  }

  for (const [key, values] of Object.entries(tests)) {
    for (const type of Object.keys(values)) {
      for (const value of values[type]) {
        createTestCase(key, value, type);
      }
    }
  }
});
