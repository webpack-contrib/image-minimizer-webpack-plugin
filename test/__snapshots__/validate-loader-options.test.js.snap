// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`validate loader options should throw an error on the "generator" option with "[]" value 1`] = `
"Invalid options object. Image Minimizer Plugin Loader has been initialized using an options object that does not match the API schema.
 - options.generator should be a non-empty array.
   -> Allows you to setup the generator function and options.
   -> Read more at https://github.com/webpack-contrib/image-minimizer-webpack-plugin#generator"
`;

exports[`validate loader options should throw an error on the "generator" option with "[{"preset":"webp","filename":true}]" value 1`] = `
"Invalid options object. Image Minimizer Plugin Loader has been initialized using an options object that does not match the API schema.
 - options.generator[0].filename should be one of these:
   non-empty string | function
   -> Allows to set the filename for the minimized asset.
   Details:
    * options.generator[0].filename should be a non-empty string.
    * options.generator[0].filename should be an instance of function."
`;

exports[`validate loader options should throw an error on the "generator" option with "[{"preset":"webp","filter":true}]" value 1`] = `
"Invalid options object. Image Minimizer Plugin Loader has been initialized using an options object that does not match the API schema.
 - options.generator[0].filter should be an instance of function.
   -> Allows filtering of images."
`;

exports[`validate loader options should throw an error on the "generator" option with "[{"preset":"webp"}]" value 1`] = `
"Invalid options object. Image Minimizer Plugin Loader has been initialized using an options object that does not match the API schema.
 - options.generator[0] misses the property 'implementation'. Should be:
   function
   -> Implementation of the generator function."
`;

exports[`validate loader options should throw an error on the "generator" option with "1" value 1`] = `
"Invalid options object. Image Minimizer Plugin Loader has been initialized using an options object that does not match the API schema.
 - options.generator should be an array:
   [object { implementation, preset, type?, options?, filter?, filename? }, ...] (should not have fewer than 1 item)
   -> Allows you to setup the generator function and options.
   -> Read more at https://github.com/webpack-contrib/image-minimizer-webpack-plugin#generator"
`;

exports[`validate loader options should throw an error on the "generator" option with "false" value 1`] = `
"Invalid options object. Image Minimizer Plugin Loader has been initialized using an options object that does not match the API schema.
 - options.generator should be an array:
   [object { implementation, preset, type?, options?, filter?, filename? }, ...] (should not have fewer than 1 item)
   -> Allows you to setup the generator function and options.
   -> Read more at https://github.com/webpack-contrib/image-minimizer-webpack-plugin#generator"
`;

exports[`validate loader options should throw an error on the "generator" option with "null" value 1`] = `
"Invalid options object. Image Minimizer Plugin Loader has been initialized using an options object that does not match the API schema.
 - options.generator should be an array:
   [object { implementation, preset, type?, options?, filter?, filename? }, ...] (should not have fewer than 1 item)
   -> Allows you to setup the generator function and options.
   -> Read more at https://github.com/webpack-contrib/image-minimizer-webpack-plugin#generator"
`;

exports[`validate loader options should throw an error on the "generator" option with "true" value 1`] = `
"Invalid options object. Image Minimizer Plugin Loader has been initialized using an options object that does not match the API schema.
 - options.generator should be an array:
   [object { implementation, preset, type?, options?, filter?, filename? }, ...] (should not have fewer than 1 item)
   -> Allows you to setup the generator function and options.
   -> Read more at https://github.com/webpack-contrib/image-minimizer-webpack-plugin#generator"
`;

exports[`validate loader options should throw an error on the "minimizer" option with "[]" value 1`] = `
"Invalid options object. Image Minimizer Plugin Loader has been initialized using an options object that does not match the API schema.
 - options.minimizer should be a non-empty array."
`;

exports[`validate loader options should throw an error on the "minimizer" option with "{"filename":true}" value 1`] = `
"Invalid options object. Image Minimizer Plugin Loader has been initialized using an options object that does not match the API schema.
 - options.minimizer should be one of these:
   [object { implementation?, options?, filter?, filename? }, ...] (should not have fewer than 1 item) | object { implementation?, options?, filter?, filename? }
   -> Allows you to setup the minimizer function and options.
   -> Read more at https://github.com/webpack-contrib/image-minimizer-webpack-plugin#minimizer
   Details:
    * options.minimizer.filename should be one of these:
      non-empty string | function
      -> Allows to set the filename for the minimized asset.
      Details:
       * options.minimizer.filename should be a non-empty string.
       * options.minimizer.filename should be an instance of function."
`;

exports[`validate loader options should throw an error on the "minimizer" option with "{"filter":true}" value 1`] = `
"Invalid options object. Image Minimizer Plugin Loader has been initialized using an options object that does not match the API schema.
 - options.minimizer.filter should be an instance of function.
   -> Allows filtering of images."
`;

exports[`validate loader options should throw an error on the "minimizer" option with "1" value 1`] = `
"Invalid options object. Image Minimizer Plugin Loader has been initialized using an options object that does not match the API schema.
 - options.minimizer should be one of these:
   [object { implementation?, options?, filter?, filename? }, ...] (should not have fewer than 1 item) | object { implementation?, options?, filter?, filename? }
   -> Allows you to setup the minimizer function and options.
   -> Read more at https://github.com/webpack-contrib/image-minimizer-webpack-plugin#minimizer
   Details:
    * options.minimizer should be an array:
      [object { implementation?, options?, filter?, filename? }, ...] (should not have fewer than 1 item)
    * options.minimizer should be an object:
      object { implementation?, options?, filter?, filename? }"
`;

exports[`validate loader options should throw an error on the "minimizer" option with "false" value 1`] = `
"Invalid options object. Image Minimizer Plugin Loader has been initialized using an options object that does not match the API schema.
 - options.minimizer should be one of these:
   [object { implementation?, options?, filter?, filename? }, ...] (should not have fewer than 1 item) | object { implementation?, options?, filter?, filename? }
   -> Allows you to setup the minimizer function and options.
   -> Read more at https://github.com/webpack-contrib/image-minimizer-webpack-plugin#minimizer
   Details:
    * options.minimizer should be an array:
      [object { implementation?, options?, filter?, filename? }, ...] (should not have fewer than 1 item)
    * options.minimizer should be an object:
      object { implementation?, options?, filter?, filename? }"
`;

exports[`validate loader options should throw an error on the "minimizer" option with "null" value 1`] = `
"Invalid options object. Image Minimizer Plugin Loader has been initialized using an options object that does not match the API schema.
 - options.minimizer should be one of these:
   [object { implementation?, options?, filter?, filename? }, ...] (should not have fewer than 1 item) | object { implementation?, options?, filter?, filename? }
   -> Allows you to setup the minimizer function and options.
   -> Read more at https://github.com/webpack-contrib/image-minimizer-webpack-plugin#minimizer
   Details:
    * options.minimizer should be an array:
      [object { implementation?, options?, filter?, filename? }, ...] (should not have fewer than 1 item)
    * options.minimizer should be an object:
      object { implementation?, options?, filter?, filename? }"
`;

exports[`validate loader options should throw an error on the "minimizer" option with "true" value 1`] = `
"Invalid options object. Image Minimizer Plugin Loader has been initialized using an options object that does not match the API schema.
 - options.minimizer should be one of these:
   [object { implementation?, options?, filter?, filename? }, ...] (should not have fewer than 1 item) | object { implementation?, options?, filter?, filename? }
   -> Allows you to setup the minimizer function and options.
   -> Read more at https://github.com/webpack-contrib/image-minimizer-webpack-plugin#minimizer
   Details:
    * options.minimizer should be an array:
      [object { implementation?, options?, filter?, filename? }, ...] (should not have fewer than 1 item)
    * options.minimizer should be an object:
      object { implementation?, options?, filter?, filename? }"
`;

exports[`validate loader options should throw an error on the "severityError" option with "() => {}" value 1`] = `
"Invalid options object. Image Minimizer Plugin Loader has been initialized using an options object that does not match the API schema.
 - options.severityError should be one of these:
   \\"off\\" | \\"warning\\" | \\"error\\"
   -> Allows to choose how errors are displayed.
   -> Read more at https://github.com/webpack-contrib/image-minimizer-webpack-plugin#severityerror"
`;

exports[`validate loader options should throw an error on the "severityError" option with "[]" value 1`] = `
"Invalid options object. Image Minimizer Plugin Loader has been initialized using an options object that does not match the API schema.
 - options.severityError should be one of these:
   \\"off\\" | \\"warning\\" | \\"error\\"
   -> Allows to choose how errors are displayed.
   -> Read more at https://github.com/webpack-contrib/image-minimizer-webpack-plugin#severityerror"
`;

exports[`validate loader options should throw an error on the "severityError" option with "{}" value 1`] = `
"Invalid options object. Image Minimizer Plugin Loader has been initialized using an options object that does not match the API schema.
 - options.severityError should be one of these:
   \\"off\\" | \\"warning\\" | \\"error\\"
   -> Allows to choose how errors are displayed.
   -> Read more at https://github.com/webpack-contrib/image-minimizer-webpack-plugin#severityerror"
`;

exports[`validate loader options should throw an error on the "severityError" option with "false" value 1`] = `
"Invalid options object. Image Minimizer Plugin Loader has been initialized using an options object that does not match the API schema.
 - options.severityError should be one of these:
   \\"off\\" | \\"warning\\" | \\"error\\"
   -> Allows to choose how errors are displayed.
   -> Read more at https://github.com/webpack-contrib/image-minimizer-webpack-plugin#severityerror"
`;

exports[`validate loader options should throw an error on the "severityError" option with "true" value 1`] = `
"Invalid options object. Image Minimizer Plugin Loader has been initialized using an options object that does not match the API schema.
 - options.severityError should be one of these:
   \\"off\\" | \\"warning\\" | \\"error\\"
   -> Allows to choose how errors are displayed.
   -> Read more at https://github.com/webpack-contrib/image-minimizer-webpack-plugin#severityerror"
`;

exports[`validate loader options should throw an error on the "unknown" option with "/test/" value 1`] = `
"Invalid options object. Image Minimizer Plugin Loader has been initialized using an options object that does not match the API schema.
 - options has an unknown property 'unknown'. These properties are valid:
   object { minimizer?, generator?, severityError? }"
`;

exports[`validate loader options should throw an error on the "unknown" option with "[]" value 1`] = `
"Invalid options object. Image Minimizer Plugin Loader has been initialized using an options object that does not match the API schema.
 - options has an unknown property 'unknown'. These properties are valid:
   object { minimizer?, generator?, severityError? }"
`;

exports[`validate loader options should throw an error on the "unknown" option with "{"foo":"bar"}" value 1`] = `
"Invalid options object. Image Minimizer Plugin Loader has been initialized using an options object that does not match the API schema.
 - options has an unknown property 'unknown'. These properties are valid:
   object { minimizer?, generator?, severityError? }"
`;

exports[`validate loader options should throw an error on the "unknown" option with "{}" value 1`] = `
"Invalid options object. Image Minimizer Plugin Loader has been initialized using an options object that does not match the API schema.
 - options has an unknown property 'unknown'. These properties are valid:
   object { minimizer?, generator?, severityError? }"
`;

exports[`validate loader options should throw an error on the "unknown" option with "1" value 1`] = `
"Invalid options object. Image Minimizer Plugin Loader has been initialized using an options object that does not match the API schema.
 - options has an unknown property 'unknown'. These properties are valid:
   object { minimizer?, generator?, severityError? }"
`;

exports[`validate loader options should throw an error on the "unknown" option with "false" value 1`] = `
"Invalid options object. Image Minimizer Plugin Loader has been initialized using an options object that does not match the API schema.
 - options has an unknown property 'unknown'. These properties are valid:
   object { minimizer?, generator?, severityError? }"
`;

exports[`validate loader options should throw an error on the "unknown" option with "test" value 1`] = `
"Invalid options object. Image Minimizer Plugin Loader has been initialized using an options object that does not match the API schema.
 - options has an unknown property 'unknown'. These properties are valid:
   object { minimizer?, generator?, severityError? }"
`;

exports[`validate loader options should throw an error on the "unknown" option with "true" value 1`] = `
"Invalid options object. Image Minimizer Plugin Loader has been initialized using an options object that does not match the API schema.
 - options has an unknown property 'unknown'. These properties are valid:
   object { minimizer?, generator?, severityError? }"
`;
